python/cudf/cudf/_libxx/scalar.pyx

```diff
@@ -0,0 +1,220 @@
+import numpy as np
+import pandas as pd
+
+from libc.stdint cimport int8_t, int16_t, int32_t, int64_t
+from libcpp.memory cimport unique_ptr
+from libcpp cimport bool
+from cudf._libxx.lib cimport *
+from cudf._libxx.lib import *
+
+
+cdef class Scalar:
+
+    cdef unique_ptr[scalar] c_value
+
+    def __init__(self, value, dtype=None):
+        """
+        cudf.Scalar: Type representing a scalar value on the device
+
+        Parameters
+        ----------
+        value : scalar
+            An object of scalar type, i.e., one for which
+            `np.isscalar()` returns `True`. Can also be `None`,
+            to represent a "null" scalar. In this case,
+            dtype *must* be provided.
+        dtype : dtype
+            A NumPy dtype.
+        """
+        from cudf.utils.dtypes import to_cudf_compatible_scalar
+
+        value = to_cudf_compatible_scalar(value, dtype=dtype)
+        valid = value is not None
+
+        if dtype is None:
```

Comment:
Would this be more pythonic?
```
try:
    dtype = np.dtype(dtype if dtype else value.dtype)
except:
    raise TypeError("dtype required when constructing a null scalar")
```

Reply:
A catch-all `except:` will get us in trouble eventually. If someone passed in `dtype='blah'`, then they'd get the error `"dtype required when constructing a null scalar"` which is confusing.

Maybe we could catch the `AttributeError` thrown on `value.dtype` when `value is None`. It definitely leads to more succinct code, but I don't know if it's easy to make sense of:

```python
try:
    dtype = np.dtype(dtype if dtype else value.dtype)
except AttributeError:
    raise TypeError("dtype required when constructing a null scalar")
```

Reply:
I think if you made that: `dtype = np.dtype(dtype if dtype is not None else value.dtype)` it's pretty clear without the ambiguity of converting a numpy dtype to a bool :)

Reply:
I'm going to merge this, but if someone feels strongly enough to change it they can submit a PR :)